plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -624,6 +628,33 @@ public ColumnMetadata getColumnMetadata(ConnectorSession session, ConnectorTable
                 .build();
     }
 
+    @Override
+    public void validateScan(ConnectorSession session, ConnectorTableHandle handle)
+    {
+        IcebergTableHandle table = (IcebergTableHandle) handle;
+        Optional<PartitionSpec> partitionSpec = table.getPartitionSpecJson()
+                .map(partitionSpecJson -> PartitionSpecParser.fromJson(SchemaParser.fromJson(table.getTableSchemaJson()), partitionSpecJson));
+        if (partitionSpec.isEmpty() || partitionSpec.get().isUnpartitioned()) {
+            return;
+        }
+        if (isQueryPartitionFilterRequired(session)
+                && table.getEnforcedPredicate().equals(TupleDomain.all())) {
```

Comment:
@zhangminglei Can you explain why this 
`table.getEnforcedPredicate().equals(TupleDomain.all())` is needed? Doesn't this mean that we only enter the block if the table has no enforced predicates?

Reply:
That's right! @marton-bod , If a table has no enforced predicates, like `select * from T where udf(partition_field) = someValue` then enter the block.

Reply:
The reason for the difference, Marton, is that some filters show up as `TupleDomain`s and some show up as `Predicates` depending on how they can be represented. 

`WHERE x = 42` for example, is a tuple domain and will show up in the enforced constraint
`WHERE x % 2 = 0` will be a Predicate because it cannot be represented as a discrete set or range

but if x is a partition column, either should be fine for this feature